File: cb-master-interview-questions-for-faang/01-arrays-and-strings/47-closest-tree-sum/47-closest-tree-sum.cpp
#include<bits/stdc++.h>
using namespace std;


int threeSumClosest(vector<int>& num, int target) {
  sort(num.begin(),num.end());
  
  int closeDiff=INT_MAX;
  int close_sum;
  
  for(int i = 0; i < num.size()-2; i++)
  {
    int sum=0;
    sum+=num[i];
  
    int j=i+1,k=num.size()-1;
    sum+=num[j]+num[k];
  
    while(j < k)
    {
      if(abs(sum-target) < closeDiff)
      {
        closeDiff = abs(sum-target);
        close_sum = sum;
      }
      if(sum < target)
      {
        sum -= num[j];
        j++;
        sum += num[j];
      }
      else if(sum > target)
      {
        sum -= num[k];
        k--;
        sum += num[k];
      }
      else
      {
        break;
      }
    }
  }
  return close_sum;
}


int main()
{
  int n;
  cin>>n;

  int target;
  cin>>target;
  
  vector<int>num;
  int a;
  
  for(int i=0;i<n;i++)
  {
    cin>>a;
    num.push_back(a);
  }
  cout<<threeSumClosest(num,target);
}

File: cb-master-interview-questions-for-faang/01-arrays-and-strings/47-closest-tree-sum/47-closest-tree-sum.py
def threeSumClosest(nums, target):
  nums.sort()
  length = len(nums)
  ans = nums[0]+nums[1]+nums[length-1]
  for i in range(length-2):
    start = i+1
    end = length-1
    while start < end:
      sum = nums[i]+nums[start]+nums[end]
      if sum > target:
        end -= 1
      else:
        start += 1
      if abs(sum-target) < abs(ans-target):
        ans = sum
  return ans

n = int(input())
target = int(input())
arr = []
for i in range(n):
  arr.append(int(input()))

res = threeSumClosest(arr,target)
print(res)